core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -923,7 +974,39 @@ public InstallState getInstallState() {
      */
     @Restricted(NoExternalUse.class)
     public void setInstallState(@Nonnull InstallState newState) {
+        if(newState.equals(installState)) {
+            return; // no change, nothing to do
+        }
+        if(InstallState.INITIAL_SETUP_COMPLETED.equals(installState)) {
+            throw new IllegalArgumentException("Setup has already been completed!");
+        }
         installState = newState;
+        try {
+            switch(newState) {
+            case INITIAL_PLUGINS_INSTALLED:
+                setInstallState(InstallState.CONFIGURING_SECURITY);
+                break;
+            case SECURITY_CONFIGURATION_COMPLETED:
+                // Do we need to create a user?
+                if(!HudsonPrivateSecurityRealm.requiresLocalUser(this)) {
+                    setInstallState(InstallState.INITIAL_SETUP_COMPLETED);
+                }
+                else {
+                    setInstallState(InstallState.CREATING_ADMIN_USER);
+                }
+                break;
```

Comment:
Again, seems to me like this should actually happen in the code that configured the security.